File: AtCoder/ACP/BootCamp4b/Medium100_2/python/C_Switches.py
N, M = map(int, input().split())
S = [list(map(lambda x: int(x) - 1, input().split()))[1:] for _ in range(M)]
P = list(map(int, input().split()))


res = 0
for i in range(1 << N):
    cnt = 0
    C = [0] * M
    for j in range(M):
        for s in S[j]:
            if i >> s & 1:
                C[j] += 1
        if C[j] % 2 == P[j]:
            cnt += 1
    if cnt == M:
        res += 1
print(res)

File: AtCoder/ACP/BootCamp4b/Medium100_2/cpp/C_Switches.cpp
#include <bits/stdc++.h>
#define MAX_M 10
using namespace std;
typedef long long ll;

ll N, M;
vector<ll> S[MAX_M];
ll P[MAX_M];
ll C[MAX_M];

void solve() {
  ll res = 0;
  for (int i = 0; i < 1 << N; ++i) {
    ll cnt = 0;
    fill(C, C + N, 0LL);
    for (int j = 0; j < M; ++j) {
      for (auto s : S[j]) {
        if (i >> s & 1) {
          ++C[j];
        }
      }
      if (C[j] % 2 == P[j]) {
        ++cnt;
      }
    }
    if (cnt == M) {
      ++res;
    }
  }
  printf("%lld\n", res);
}

int main() {
  cin >> N >> M;
  for (int i = 0; i < M; ++i) {
    ll k;
    cin >> k;
    S[i] = vector<ll>(k);
    for (int j = 0; j < k; ++j) {
      cin >> S[i][j];
      --S[i][j];
    }
  }
  for (int i = 0; i < M; ++i) {
    cin >> P[i];
  }
  solve();
}